plaidml/bridge/openvino/ops/prior_box.cpp

```diff
@@ -0,0 +1,444 @@
+  void processMinSizePath();
+
+  // Append variance after former result
+  void processVariance();
+
+  edsl::Tensor& getOutput() { return C_out; }
+
+ private:
+  // Preprocess data with some attributes
+  void prepareConfig();
+
+  // Produce a tensor which data are interlaced with W and H
+  // Use this to produce box center for different path
+  void produceCenterBase();
+
+  const Context& ctx;
+  const ngraph::op::PriorBoxAttrs& attrs;
+  int H;
+  int W;
+  int IH;
+  int IW;
+  int num_of_prior;
+  int min_element_size;
+  float step;
+  float step_x;
+  float step_y;
+  float offset;
+  std::vector<float> aspect_ratios;
+  std::vector<float> aspect_ratios_scale_size;
+  std::vector<float> variance;
+  std::vector<float> min_size;
+  edsl::Tensor C_mix;
+  edsl::Tensor C_out;
+};
+
+PriorBoxImpl::PriorBoxImpl(const Context& context, const ngraph::op::PriorBoxAttrs& attributes)
+    : ctx(context), attrs(attributes) {
+  prepareConfig();
+  produceCenterBase();
+}
+
+// Preprocess data with some attributes
+void PriorBoxImpl::prepareConfig() {
+  auto* input_shape_ngraph_op = ngraph::as_type<ngraph::op::Constant>(ctx.layer->get_input_node_ptr(0));
+  std::vector<int> input_shape = input_shape_ngraph_op->get_vector<int>();
+  H = input_shape[0];
+  W = input_shape[1];
+  auto* image_shape_ngraph_op = ngraph::as_type<ngraph::op::Constant>(ctx.layer->get_input_node_ptr(1));
+  std::vector<int> image_shape = image_shape_ngraph_op->get_vector<int>();
+  IH = image_shape[0];
+  IW = image_shape[1];
+
+  num_of_prior = ngraph::op::PriorBox::number_of_priors(attrs);
+
+  aspect_ratios.push_back(1.0f);
+  for (const auto& aspect_ratio : attrs.aspect_ratio) {
+    bool exist = false;
+    for (const auto existed_value : aspect_ratios) exist |= std::fabs(aspect_ratio - existed_value) < 1e-6;
+
+    if (!exist) {
+      aspect_ratios.push_back(aspect_ratio);
+      if (attrs.flip) {
+        aspect_ratios.push_back(1.0f / aspect_ratio);
+      }
+    }
+  }
+  for (float ar : aspect_ratios) {
+    if (std::fabs(ar - 1.0f) < 1e-6) {
+      continue;
+    }
+    aspect_ratios_scale_size.push_back(std::sqrt(ar));
+  }
+
+  variance = attrs.variance;
+  if (variance.empty()) variance.push_back(0.1f);
+
+  step = attrs.step;
+  min_size = attrs.min_size;
+  if (!attrs.scale_all_sizes) {
+    // mxnet-like PriorBox
+    if (step == -1)
+      step = 1.f * IH / H;
+    else
+      step *= IH;
+    for (auto& size : min_size) size *= IH;
+  }
+  step_x = (step == 0) ? static_cast<float>(IW) / W : step;
+  step_y = (step == 0) ? static_cast<float>(IH) / H : step;
+
+  min_element_size = min_size.size();
+  offset = attrs.offset;
+}
+
+void PriorBoxImpl::produceCenterBase() {
+  auto CW = edsl::cast(edsl::index({edsl::TensorDim(H), edsl::TensorDim(W), edsl::TensorDim(1)}, 1), DType::FLOAT32);
+  auto CH = edsl::cast(edsl::index({edsl::TensorDim(H), edsl::TensorDim(W), edsl::TensorDim(1)}, 0), DType::FLOAT32);
```

Comment:
I notice the use of `FLOAT32`, but I wonder how this will effect networks that are intended to run with FLOAT16. Is there some other data type we can use to cast into rather than hard coding the type?

Reply:
For precision issues, just follow the reference implementation as now the output will be compared with it, i think it make sense  to use different precision if we don't  need to compare with it which needs to force set RefMode to  IE mode

Reply:
I believe in all places `FLOAT32` is currently used, it would be correct to replace with the requested output type. And that should be available as `ctx.layer->get_element_type()` since PriorBox is a single-output op.

Reply:
Yes, shall do this, i use get_output_element_type(0) to get the type as the comments on get_element_type() advice